src/scheduler/pbs_sched.c

```diff
@@ -792,45 +506,400 @@ are_we_primary()
+				   msg_daemonname, "Server rejected register request for %s connection with error: %s",
+				   is_primary ? "primary" : "secondary", errmsg);
+			die(-1);
+		}
+		goto rerr;
+	}
+
+	PBSD_FreeReply(reply);
+	return sock;
+
+rerr:
+	pbs_disconnect(sock);
+	PBSD_FreeReply(reply);
+	return -1;
+}
+
+/**
+ * @brief connect to given server, create and add server struct in servers list
+ *        also add secondary connection to poll list
+ *
+ * @param[in] svrhost - address of server in host[:port] format
+ *
+ * @return void
  */
+static void
+connect_server(char *svrhost)
+{
+	int primary_sock = -1;
+	int secondary_sock = -1;
+	sched_svrconn *svr;
+	void *svrs_newarr;
+
+	if (svrhost == NULL)
+		return;
+
+	while (primary_sock < 0 || secondary_sock < 0) {
+		if (primary_sock < 0) {
+			primary_sock = connect_server_helper(svrhost, 1);
+			if (primary_sock < 0) {
+				/* wait for 2s for not to burn too much CPU, and then retry connection */
+				sleep(2);
+				continue;
+			}
+		}
+		secondary_sock = connect_server_helper(svrhost, 0);
+		if (secondary_sock < 0) {
+			/* wait for 2s for not to burn too much CPU, and then retry connection */
+			sleep(2);
+			continue;
+		}
+	}
+	if (tpp_em_add_fd(poll_context, secondary_sock, EM_IN | EM_HUP | EM_ERR) < 0) {
+		log_errf(errno, __func__, "Couldn't add secondary connection to poll list for server %s", svrhost);
+		pbs_disconnect(primary_sock);
+		pbs_disconnect(secondary_sock);
+		die(-1);
+	}
+
+	svr = malloc(sizeof(sched_svrconn));
+	if (!svr) {
+		log_err(errno, __func__, MEM_ERR_MSG);
+		die(-1);
+	}
+
+	svr->svrhost = strdup(svrhost);
+	if (!svr->svrhost) {
+		free(svr);
+		log_err(errno, __func__, MEM_ERR_MSG);
+		die(-1);
+	}
+	svr->primary_sock = primary_sock;
+	svr->secondary_sock = secondary_sock;
+	svrs_newarr = add_ptr_to_array(servers, svr);
+	if (svrs_newarr == NULL) {
+		close_server(svr);
+		die(-1);
+	}
+	servers = (sched_svrconn **)svrs_newarr;
+	log_eventf(PBSEVENT_ADMIN | PBSEVENT_FORCE, PBS_EVENTCLASS_SCHED, LOG_INFO, msg_daemonname, "Connected to server %s", svrhost);
+}
+
+/**
+ * @brief Connect to all configured servers and populate servers list with
+ *        connections to servers
+ *
+ * @return void
+ */
+static void
+connect_servers(void)
+{
+	int max_connection = 1;
+	char *svrhost = pbs_default();
+
+	if (poll_context == NULL) {
+		poll_context = tpp_em_init(max_connection);
+		if (poll_context == NULL) {
+			log_err(errno, __func__, "Failed to init cmd connections context");
+			die(-1);
+		}
+
+		sched_cmds = new_ds_queue();
+		if (sched_cmds == NULL) {
+			log_errf(errno, __func__, MEM_ERR_MSG);
+			die(-1);
+		}
+	}
+
+	connect_server(svrhost ? svrhost : "");
+}
+
+/**
+ * @brief reconnect to given server
+ *
+ * @param[in] sconn - pointer server struct to reconnect
+ *
+ * @return void
+ */
+static void
+reconnect_server(sched_svrconn *sconn)
+{
+	char svrhost[PBS_MAXSERVERNAME + PBS_MAXPORTNUM + 2];
+
+	strcpy(svrhost, sconn->svrhost);
```

Comment:
please use strncpy.

Reply:
I don't think we need it. Because, sconn->svrhost comes from pbs.conf and max value of it what I have used as a sizeof svrhost[] above. So strcpy will be suffice.

Reply:
I'm not quite sure I understand.  Are you saying that when we read out of pbs.conf, we do the truncation of a very long host there?

Reply:
Yes, see [this](https://github.com/openpbs/openpbs/blob/master/src/lib/Libifl/pbsD_connect.c#L107) for single server and for multi-server see [this](https://github.com/openpbs/openpbs/blob/master/src/lib/Libifl/pbs_loadconf.c#L295) and [this](https://github.com/openpbs/openpbs/blob/master/src/lib/Libutil/misc_utils.c#L2329)